sa/sa.go

```diff
@@ -2132,3 +2133,46 @@ func (ssa *SQLStorageAuthority) GetAuthz2(id int64) (*corepb.Authorization, erro
 	}
 	return modelToAuthzPB(obj.(*authz2Model))
 }
+
+// RevokeCertificate stores revocation information about a certificate. It will only store this
+// information if the certificate is not alreay marked as revoked. This method is meant as a
+// replacement for MarkCertificateRevoked and the ocsp-updater database methods.
+func (ssa *SQLStorageAuthority) RevokeCertificate(ctx context.Context, req *sapb.RevokeCertificateRequest) error {
+	tx, err := ssa.dbMap.Begin()
+	if err != nil {
+		return err
+	}
+	txWithCtx := tx.WithContext(ctx)
+
+	status, err := SelectCertificateStatus(
+		txWithCtx,
+		"WHERE serial = ? AND status != ?",
+		*req.Serial,
+		string(core.OCSPStatusRevoked),
+	)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			// InternalServerError because we expected this certificate status to exist and
+			// not be revoked.
+			return Rollback(tx, berrors.InternalServerError("no certificate with serial %s", *req.Serial))
```

Comment:
I think this error message might confuse an operator when the certificate with the listed serial does exist but isn't the expected status. 
```suggestion
			return Rollback(tx, berrors.InternalServerError("no certificate with serial %s and status %s", *req.Serial, string(core.OCSPStatusRevoked)))
```